pkg/controller/installation/products/codeready/reconciler.go

```diff
@@ -0,0 +1,385 @@
+	r.logger.Infof("checking installplan is created for subscription %s in namespace: %s", defaultSubscriptionName, r.Config.GetNamespace())
+	ip, err := r.mpm.GetSubscriptionInstallPlan("codeready-workspaces", r.Config.GetNamespace())
+	if err != nil {
+		if k8serr.IsNotFound(err) {
+			r.logger.Infof(fmt.Sprintf("installplan resource is not found in namespace: %s", r.Config.GetNamespace()))
+			return v1alpha1.PhaseAwaitingOperator, nil
+		}
+		return v1alpha1.PhaseFailed, pkgerr.Wrap(err, fmt.Sprintf("could not retrieve installplan in namespace: %s", r.Config.GetNamespace()))
+	}
+
+	r.logger.Infof("installplan phase is %s, waiting for %s", ip.Status.Phase, coreosv1alpha1.InstallPlanPhaseComplete)
+	if ip.Status.Phase != coreosv1alpha1.InstallPlanPhaseComplete {
+		return v1alpha1.PhaseAwaitingOperator, nil
+	}
+	return v1alpha1.PhaseCreatingComponents, nil
+}
+
+func (r *Reconciler) handleProgressPhase(ctx context.Context) (v1alpha1.StatusPhase, error) {
+	r.logger.Infof("checking pod status in namespace: %s", r.Config.GetNamespace())
+	pods, err := r.coreClient.CoreV1().Pods(r.Config.GetNamespace()).List(metav1.ListOptions{})
+	if err != nil {
+		return v1alpha1.PhaseFailed, pkgerr.Wrap(err, fmt.Sprintf("failed to retrieve pods in namespace: %s", r.Config.GetNamespace()))
+	}
+
+	//expecting 3 pods in total
+	if len(pods.Items) < 3 {
+		return v1alpha1.PhaseInProgress, nil
+	}
+
+	//and they should all be ready
+checkPodStatus:
+	for _, pod := range pods.Items {
+		for _, cnd := range pod.Status.Conditions {
+			if cnd.Type == v1.ContainersReady {
+				if cnd.Status != v1.ConditionStatus("True") {
+					r.logger.Infof("pod not ready, returning in progress: %+v", cnd.Status)
+					return v1alpha1.PhaseInProgress, nil
+				}
+				break checkPodStatus
+			}
+		}
+	}
+
+	return v1alpha1.PhaseCompleted, nil
+}
+
+func (r *Reconciler) reconcileKeycloakClient(ctx context.Context) (v1alpha1.StatusPhase, error) {
+	r.logger.Infof("checking keycloak client exists in keycloakrealm custom resource in namespace: %s", r.Config.GetNamespace())
+	serverClient, err := pkgclient.New(r.restConfig, pkgclient.Options{})
+	if err != nil {
+		return v1alpha1.PhaseFailed, err
+	}
+
+	// retrive the checluster so we can use its URL for redirect and web origins in the keycloak client
+	cheCluster := &chev1.CheCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      defaultCheClusterName,
+			Namespace: r.Config.GetNamespace(),
+		},
+	}
+	serverClient.Get(ctx, pkgclient.ObjectKey{Name: defaultCheClusterName, Namespace: r.Config.GetNamespace()}, cheCluster)
+	cheURL := cheCluster.Status.CheURL
+	if cheURL == "" {
+		return v1alpha1.PhaseFailed, errors.New("che URL is not set")
+	}
+
+	// retrieve the sso config so we can find the keycloakrealm custom resource to update
+	kcCfg, err := r.ConfigManager.ReadRHSSO()
+	if err != nil {
+		return v1alpha1.PhaseFailed, err
+	}
+	if kcCfg.GetRealm() == "" {
```

Comment:
I think we  could prob avoid a lot if this logic if we had a ```cfg.Valid() ``` func that is called at start up

Reply:
Yeah this makes sense